### PR TITLE
Make Agent a governed drop-in chat

### LINKED ISSUE #33: Report the compaction method and allow opting out of the degraded fallback
## Problem

When LLM summarisation fails, compaction falls back to a truncated text summary. That fallback is now visible — it warns and emits a `compact_fallback` notification — but two gaps remain:

1. **No way to tell after the fact.** Nothing on the result records which method produced the summary, so a caller inspecting a compacted conversation cannot tell whether it holds a real summary or truncated text.
2. **No way to refuse.** Some callers would rather fail than continue with badly degraded context. There is no option to say so.

## Where

`Agent$compact()` and the private `generate_compaction_summary()` in `R/agent.R`

## Done when

- [ ] The compaction method used is recorded where a caller can read it
- [ ] Callers can choose between falling back and failing

### LINKED ISSUE #46: Automate context compaction and large-result offload
## Problem

Compaction is manual. A long run grows context until the provider rejects it, and the caller is expected to notice and call `compact()` first.

## Scope

- Configurable context thresholds that trigger compaction automatically
- Summarisation boundaries chosen deliberately rather than by turn count
- PreCompact and PostCompact lifecycle events around automatic compaction, as for manual
- Durable offload for large tool results, with references left in context
- Preserved provenance and budget accounting across both

## Where

`Agent$compact()` and the run loop in `R/agent.R`

## Done when

- [ ] A long run compacts without caller intervention
- [ ] Large tool results are offloaded with a resolvable reference
- [ ] Automatic compaction fires the same lifecycle events as manual

Depends on knowing which compaction method produced a summary — see the compaction reporting issue.

### LINKED ISSUE #63: LeadAgent delegation blocks the R process under run_shiny()
## Problem

`LeadAgent` inherits `run_shiny()` from `Agent`, so it looks usable in a Shiny app. Its delegation path is synchronous:

```r
sub_result <- sub_agent$run_sync(   # R/agents-multi.R:493
```

`run_shiny()` is *cooperative* async — a `coro::async_generator` driven by ellmer's `stream_async()`, running on the main R thread. A blocking `run_sync()` inside a tool call does not yield. It holds the thread for the entire subagent run, which is a full multi-turn agentic loop with its own tool calls.

## Impact

In a single-process Shiny app, one delegation freezes **every connected session**, not just the one that triggered it. The UI stops streaming, other users' inputs queue, and nothing indicates why. The longer the subagent's task, the longer the freeze.

This is worse than a missing feature: the API advertises Shiny support on a class where using it degrades the whole app.

## Where

- `run_shiny()` in `R/agent.R`
- The delegation call in `R/agents-multi.R`

## Options

1. Give `LeadAgent` an async delegation path that yields to the event loop
2. Move subagent runs off-thread (mirai/callr) — overlaps with the background subagent work
3. Short term: make `LeadAgent$run_shiny()` refuse, so the failure is loud rather than a silent freeze

Option 3 is not a fix but it is honest, and it is small.

## Done when

- [ ] Delegating under `run_shiny()` either yields to the event loop or fails loudly
- [ ] A test covers a delegation triggered from the Shiny path

## Context

Filed while establishing that most agents will run inside shinychat. Under that assumption this stops being an edge case and becomes the main path.

### LINKED ISSUE #68: No integration tests exercise shinychat, the primary host
## Problem

ADR-0002 names shinychat as the primary host for deputy agents. There is not one test in the suite that loads shinychat.

```
$ grep -rl "shinychat" tests/testthat/
(no matches)
```

`Agent$run_shiny()` and the newer `Agent$run_async()` are tested through their own machinery, but nothing verifies that what `run_shiny()` returns actually works when handed to `shinychat::chat_append()` — which is its entire stated purpose and the one thing `inst/examples/shiny-chat/app.R` does.

## Why this matters now

Upstream is moving, and deputy has no way to notice:

- `chat_mod_server()` and `chat_mod_ui()` are **soft-deprecated** in favour of `chat_server()`
- `input$<id>_user_input` **changes shape** when attachments are enabled: a plain string becomes a list of ellmer `Content` objects. `chat_server()` enables attachments by default. The example app reads `input$chat_user_input` directly and assumes a string
- The whole conversation store arrives with the next release

Each of these is a silent break waiting for a user to find.

## Scope

- A test that passes `run_shiny()` output through `chat_append()` and asserts the stream is consumed
- A test covering the attachment-enabled input shape, so the `Content`-list form is handled rather than assumed away
- Once the store lands and #66 is resolved, round-trip tests against `ConversationStore`

Tests should skip when shinychat is unavailable, in the usual way — but they must exist to be skipped.

## Done when

- [ ] `run_shiny()` is tested against a real `chat_append()`
- [ ] Both `user_input` shapes are covered
- [ ] The example app is verified against current shinychat, not the version it was written for

### LINKED ISSUE #77: Consolidate run lifecycle and workspace resolution behind one kernel
## Problem

Deputy currently has two substantial run engines:

- the callback-driven `stream_async()` path used by `run_shiny()` and `run_async()`;
- the coro generator path used by `run()` and `run_sync()`.

They separately own run state, hooks, limits, checkpoints, provider failures, stop reasons, cleanup, and results. Their public capabilities already differ. The synchronous path also changes the process-wide working directory around generator steps, while async paths require absolute file paths.

This is a divergent-change hotspot and a concurrency-sensitive workspace seam.

## Direction

Create one internal run kernel/state machine that emits a canonical event/result model. Make sync generators, promises, and Shiny projections thin adapters. Introduce an explicit workspace/path resolver consumed by native tools and permissions; remove process-global `setwd()` from the run lifecycle.

This is the architectural seam that should guide the Agent portion of #72 rather than mechanically moving methods between files.

## Done when

- [ ] One kernel owns run start, state transitions, limits, hooks, terminal accounting, and cleanup.
- [ ] `run()`, `run_sync()`, `run_async()`, and `run_shiny()` share contract tests.
- [ ] Documented capability differences are deliberate adapter constraints, not duplicated-engine drift.
- [ ] Native file tools resolve against an explicit immutable workspace.
- [ ] Agent runs no longer mutate the R process working directory.
- [ ] Cancellation and abandoned-generator cleanup remain fail closed.
- [ ] The work lands in reviewable vertical slices coordinated with #63 and #72.

### PR DESCRIPTION
**Pull request by JamesHWade**
## Summary

- replace the exposed backend and separate `run_shiny()` path with Chat-shaped methods over one governed async kernel
- add automatic context compaction, explicit fallback reporting, and durable large-result offload
- make LeadAgent delegation asynchronous and test real shinychat streaming and attachment inputs
- resolve related lifecycle, workspace, and lint debt

## Breaking changes

- remove the public `agent$chat` backend binding
- remove `Agent$run_shiny()`; use `Agent$stream_async()` directly
- stop retrying failed provider streams through a separate `chat()` request

## Backlog

Closes #33
Closes #46
Closes #63
Closes #68
Closes #77

## Verification

- `air format R/ tests/testthat/`
- `jarl check R/`
- `devtools::test()` — 2,149 passing
- `devtools::check()` — 0 errors, 0 warnings, 1 clock-verification note
- `git diff --check`

### CODE REVIEW

File: R/agent.R [hunk 1]
```diff
@@ -2968,647 +3343,129 @@ Agent <- R6::R6Class(
...
+    },
+
+    system_prompt_without_compaction = function() {
+      prompt <- private$.chat$get_system_prompt() %||% ""
+      marker <- "\n\n## Previous Conversation Summary\n"
+      location <- regexpr(marker, prompt, fixed = TRUE)
+      if (location[[1L]] < 0L) {
+        return(prompt)
+      }
+      substr(prompt, 1L, location[[1L]] - 1L)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R]: Fixed in 7e5dc0e. Compaction summaries now use explicit start/end delimiters, and replacement removes only that block, preserving prompt content appended afterward. The repeated-compaction regression verifies hook-added context survives while the old summary is replaced.

File: R/agent.R [hunk 2]
```diff
@@ -221,23 +226,23 @@ Agent <- R6::R6Class(
...
       # Override system prompt if provided
       if (!is.null(system_prompt)) {
-        self$chat$set_system_prompt(system_prompt)
+        private$.chat$set_system_prompt(system_prompt)
       }
 
       # Register tools
       if (length(tools) > 0) {
-        self$chat$register_tools(tools)
+        self$register_tools(tools)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R]: Fixed in 7e5dc0e. Agent initialization now captures tools already registered on the supplied Chat, clears the backend registry, and re-registers each tool through this Agent's runtime adapter. The regression test verifies that a preloaded relative-path write resolves inside the receiving Agent's working directory.

File: R/tool-runtime.R [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  "grep_files",
+  "read_csv"
+)
+
+runtime_wrap_tool <- function(tool, agent) {
+  if (!inherits(tool, "ellmer::ToolDef")) {
+    cli_abort("{.arg tool} must be an ellmer tool definition")
+  }
+  if (isTRUE(attr(tool, "deputy_runtime_tool"))) {
+    return(tool)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/tool-runtime.R]: Fixed in 7e5dc0e. Runtime adaptation now unwraps any prior Deputy wrapper to its source ToolDef and creates a fresh wrapper with the receiving Agent's argument resolver and result processor. The regression test copies Agent 1's registered tool into Agent 2 and verifies the write occurs only in Agent 2's workspace.

File: R/tool-runtime.R [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+  )
+}
+
+tool_result_reference_text <- function(record) {
+  paste(
+    "[Tool result offloaded by Deputy]",
+    paste0("reference: ", record$uri),
+    paste0("bytes: ", record$bytes),
+    paste0("sha256: ", record$sha256),
+    "Use Agent$resolve_tool_result(reference) to retrieve the complete value.",
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/tool-runtime.R]: Fixed in 7e5dc0e. An offloaded value now leaves a bounded preview and causes Deputy to register a governed, read-only `deputy_read_tool_result` tool. The model can page through integrity- and session-validated chunks (maximum 16,384 characters); host code can still recover the original R value. Tests cover model retrieval, tool replacement, and cloned chats.

File: R/agent.R [hunk 5]
```diff
@@ -1446,85 +1741,290 @@ Agent <- R6::R6Class(
...
+    },
+
+    ensure_tool_result_reader = function() {
+      if (isTRUE(private$.tool_result_reader_registered)) {
+        return(invisible(NULL))
+      }
+      private$.chat$register_tool(
+        private$adapt_tool(tool_result_reader_tool(
+          private$read_tool_result_chunk
+        ))
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R:1837]: Fixed in bc915ea. The session-local reader now carries a package-private marker through Agent tool adaptation. Only that marked `deputy_read_tool_result` pair bypasses allowlist exclusion; an explicit denylist still wins. The regression test exercises both sides: the internal reader is admitted under `tool_allowlist = "large_result"`, while an ordinary non-allowlisted tool remains rejected.

File: R/tool-runtime.R [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+  if (!is_absolute_path(arguments$path)) {
+    arguments$path <- file.path(working_dir, arguments$path)
+  }
+  arguments
+}
+
+tool_result_offload_dir <- function(policy, session_id) {
+  root <- policy$offload_dir %||%
+    file.path(tools::R_user_dir("deputy", "cache"), "tool-results")
+  file.path(path.expand(root), session_id)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/tool-runtime.R:107]: Fixed in bc915ea. Saved sessions now embed integrity-checked result envelopes, validate them before conversation mutation, atomically import them beneath the receiving Agent’s retained session identity, and restore the bounded reader tool. Tests cover cross-Agent recovery and atomic rejection of a tampered envelope.

File: R/agent.R [hunk 7]
```diff
@@ -3655,9 +3788,19 @@ Agent <- R6::R6Class(
       )
 
       conversation_text <- paste(turn_texts, collapse = "\n\n")
+      prior_summary <- private$.compaction_summary
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R:4061]: Fixed in bc915ea. The strict session payload now carries `compaction_summary`, and load restores it alongside the prompt. The regression test saves after compaction, loads into a fresh Agent, and verifies the next LLM compaction prompt merges the saved cumulative summary.

File: R/agent.R [hunk 8]
```diff
@@ -1969,6 +2651,8 @@ Agent <- R6::R6Class(
         "schema_version",
         "turns",
         "system_prompt",
+        "compaction_summary",
+        "tool_result_envelopes",
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R:2717]: Addressed in 4bc4797. This PR is intentionally allowed to break pre-production session files, so the expanded strict payload is now schema v2 rather than a silent incompatible change to v1. The loader checks the version before v2-only fields and reports v1 explicitly as unsupported; the regression test covers that contract.

File: R/agent.R [hunk 9]
```diff
@@ -3666,63 +3895,58 @@ Agent <- R6::R6Class(
...
-      summary
+      private$notify(
+        "Compaction used the configured deterministic text fallback.",
+        level = "warning",
+        code = "compact_fallback",
+        error = response$message
+      )
+      list(
+        summary = private$generate_fallback_summary(turns),
+        method = "text",
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R:4131]: Fixed in 4bc4797. Deterministic fallback summaries now prepend `.compaction_summary` before the newly compacted excerpt, so replacing the prompt block retains all earlier compacted context. The regression performs a custom first compaction, forces provider failure on the next one, and verifies both the prior summary and new turns survive.

File: R/tool-runtime.R [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+
+tool_result_model_text <- function(value) {
+  if (is.character(value)) {
+    return(paste(value, collapse = "\n"))
+  }
+  paste(utils::capture.output(dput(value)), collapse = "\n")
+}
+
+tool_result_preview <- function(value, max_chars = 4000L) {
+  text <- tool_result_model_text(value)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/tool-runtime.R]: Fixed in 4bc4797. Character previews are now assembled incrementally up to the character budget, and non-character values use a bounded structural rendering with capped depth, vector length, list length, string width, and output width before final truncation. Tests cover a huge character vector, nested list, and 10,000-column data frame while asserting bounded output.

File: R/agent.R [hunk 11]
```diff
@@ -1434,9 +1672,69 @@ Agent <- R6::R6Class(
...
+        }
+      }
+
+      tools <- private$.chat$get_tools()
+      had_result_reader <- "deputy_read_tool_result" %in% names(tools)
+      tools[["deputy_read_tool_result"]] <- NULL
+      private$.chat$set_tools(list())
+      if (length(tools) > 0L) {
+        private$.chat$register_tools(lapply(tools, private$adapt_tool))
+      }
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R:1742]: Fixed in 6be6e79. Clone rewiring now has a subclass hook, and LeadAgent uses it to recreate delegate_to_agent from the clone own self/private state before runtime wrapping. The regression test adds a clone-only definition and verifies it is visible only through the clone delegate tool.

File: R/agent.R [hunk 12]
```diff
@@ -150,6 +153,7 @@ Agent <- R6::R6Class(
       system_prompt = NULL,
       permissions = NULL,
       usage_limits = UsageLimits(max_requests = 25),
+      context_policy = ContextPolicy(),
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R:156]: Fixed in 6be6e79. LeadAgent$new() now exposes and documents context_policy, forwards it to Agent, and propagates the same immutable policy to delegated agents. The focused regression asserts both the lead and child receive the configured policy.

File: R/agents-multi.R [hunk 13]
```diff
@@ -621,15 +621,41 @@ LeadAgent <- R6::R6Class(
...
           }
 
-          result <- tryCatch(
-            {
-              sub_result <- sub_agent$run_sync(
-                task_to_run
-              )
+          promises::promise_resolve(NULL) |>
+            promises::then(function(...) {
+              sub_agent$run_async(task_to_run)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agents-multi.R:664]: Fixed in 6be6e79. LeadAgent now reserves every finite UsageLimits allocation before launching the child. Sibling limit derivation subtracts active reservations, and success or failure releases the reservation before charging actual child usage. Regression coverage verifies sibling budgets and settlement on both paths.

File: R/agent.R [hunk 14]
```diff
@@ -350,76 +351,308 @@ Agent <- R6::R6Class(
...
+    get_system_prompt = function() {
+      private$.chat$get_system_prompt()
+    },
+
+    #' @description Replace the system prompt, as in ellmer Chat.
+    #' @param value The new system prompt or `NULL`.
+    #' @return Invisible self.
+    set_system_prompt = function(value) {
+      private$.chat$set_system_prompt(value)
+      private$.compaction_summary <- NULL
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R]: Fixed in a64de3b. set_system_prompt() now parses a valid Deputy compaction block and restores its cumulative summary instead of clearing state unconditionally. The regression round-trips an appended prompt, performs another real compaction, and verifies the prior summary is included and the appended content survives.

File: R/tool-runtime.R [hunk 15]
```diff
@@ -0,0 +1,519 @@
...
+        tool_name,
+        arguments
+      )
+      value <- do.call(original, arguments)
+      if (promises::is.promising(value)) {
+        return(promises::then(value, function(resolved) {
+          process_result(tool_name, resolved)
+        }))
+      }
+      process_result(tool_name, value)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/tool-runtime.R]: Fixed in a64de3b. Runtime tool execution now captures a run-scoped handoff keyed to the exact tool call before invocation. The post-tool hook claims the original large value while ellmer and the model receive only the bounded reference; the handoff is removed after the callback and cleared at run teardown. The regression exercises the request, wrapped tool, and result callback path end to end.

File: R/agents-multi.R [hunk 16]
```diff
@@ -326,13 +331,13 @@ LeadAgent <- R6::R6Class(
 
       # Rebuild and update system prompt to include new sub-agent
       # Extract base prompt (before sub-agent section) if possible
-      current_prompt <- self$chat$get_system_prompt()
+      current_prompt <- private$.chat$get_system_prompt()
       base_prompt <- private$extract_base_prompt(current_prompt)
       new_prompt <- private$build_lead_prompt(
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agents-multi.R]: Fixed in a64de3b. LeadAgent routing now has explicit start/end boundaries, and register_sub_agent() replaces only that generated section. The regression compacts a lead, appends hook context, registers another agent, and verifies the summary, hook context, and single updated routing section all remain.

File: R/tool-runtime.R [hunk 17]
```diff
@@ -0,0 +1,530 @@
...
+  session_id
+) {
+  offset <- context_policy_nonnegative_whole_number(offset, "offset")
+  max_chars <- context_policy_whole_number(max_chars, "max_chars")
+  if (max_chars > 16384L) {
+    cli_abort("{.arg max_chars} must not exceed 16384")
+  }
+
+  envelope <- read_tool_result_envelope(reference, policy, session_id)
+  text <- tool_result_model_text(envelope$value)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/tool-runtime.R]: Fixed in dd0fd99. Offload now persists chunkable text plus a lightweight integrity manifest. The reference binds the text hash; each read validates the manifest and sidecar, then skips and reads in bounded blocks without loading or rendering the original R object. Session import recreates and validates the artifacts. Tests cover large character and nested values and prove chunking works while the full-value RDS is unavailable.

File: R/tool-runtime.R [hunk 18]
```diff
@@ -0,0 +1,530 @@
...
+  if (!is.null(internal_tool)) {
+    attr(wrapped, "deputy_internal_tool") <- internal_tool
+  }
+  wrapped
+}
+
+resolve_runtime_tool_arguments <- function(tool_name, arguments, working_dir) {
+  if (
+    !tool_name %in% native_file_tool_names ||
+      is.null(arguments$path) ||
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/tool-runtime.R:89]: Fixed in dd0fd99. Deputy built-in R and bash tools now carry a private workspace-runner marker. Agent runtime dispatch passes its configured working_dir only to those marked built-ins; callr receives wd explicitly and the no-callr shell fallback changes directory inside the child command, never in the host R process. Regression tests verify both tools write only inside the Agent workspace.

File: R/tool-runtime.R [hunk 19]
```diff
@@ -0,0 +1,816 @@
...
+  if (!is_absolute_path(arguments$path)) {
+    arguments$path <- file.path(working_dir, arguments$path)
+  }
+  arguments
+}
+
+tool_result_offload_dir <- function(policy, session_id) {
+  root <- policy$offload_dir %||%
+    file.path(tools::R_user_dir("deputy", "cache"), "tool-results")
+  file.path(path.expand(root), session_id)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/tool-runtime.R:107]: Fixed in 24df4c6. ContextPolicy now expands and anchors an explicit relative offload_dir against the construction-time working directory, then stores its normalized absolute path. A regression offloads in one process directory, changes the host directory, and successfully reads the same result without creating a second relative root.

File: R/agent.R [hunk 20]
```diff
@@ -2574,8 +3306,24 @@ Agent <- R6::R6Class(
...
+        identical(
+          extracted$internal_tool,
+          deputy_tool_result_reader_marker
+        ) &&
+        identical(tool_name, "deputy_read_tool_result")
+      perm_result <- if (
+        isTRUE(internal_result_reader) &&
+          !tool_name %in% self$permissions$tool_denylist
+      ) {
+        PermissionResultAllow()
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R]: Fixed in 24df4c6. The internal marker now exempts only the allowlist gate inside Permissions. Case-insensitive denylist matching still runs first, custom callback denials are honored in every mode for this reader, and the remaining mode and capability checks still apply. Regressions cover an uppercase denylist entry and an explicit callback veto.

File: R/agent.R [hunk 21]
```diff
@@ -350,76 +351,313 @@ Agent <- R6::R6Class(
...
+      }
+      private$.chat$get_turns()
+    },
+
+    #' @description Replace conversation turns, as in ellmer Chat.
+    #' @param value A list of ellmer turns.
+    #' @return Invisible self.
+    set_turns = function(value) {
+      private$.chat$set_turns(value)
+      invisible(self)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R:562]: Fixed in 24df4c6. Agent set_turns now removes only the bounded Deputy compaction block from the current system prompt and clears the cumulative compaction summary after replacing turns. Base prompt and later hook context remain intact. The regression compacts a conversation, appends context, clears turns, and verifies no historical summary survives.

File: R/agent.R [hunk 22]
```diff
@@ -350,76 +351,319 @@ Agent <- R6::R6Class(
...
+
+    #' @description Replace conversation turns, as in ellmer Chat.
+    #' @param value A list of ellmer turns.
+    #' @return Invisible self.
+    set_turns = function(value) {
+      private$.chat$set_turns(value)
+      prompt <- private$.chat$get_system_prompt()
+      parts <- private$compaction_prompt_parts(prompt)
+      if (!is.null(parts)) {
+        private$.chat$set_system_prompt(paste0(parts$before, parts$after))
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R]: Fixed in db5b46c. Compaction removal now requires active Deputy provenance: a stored cumulative summary whose content exactly matches the bounded prompt block. set_system_prompt preserves that provenance only when round-tripping the same active summary; arbitrary matching headings do not create it. Regressions cover both an existing user prompt and the public prompt setter while retaining the earlier compacted-prompt round trip.

File: R/agent.R [hunk 23]
```diff
@@ -1644,13 +2189,21 @@ Agent <- R6::R6Class(
           )
         }
 
+        compaction <- agent$.__enclos_env__$private$maybe_auto_compact(messages)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R]: Fixed in db5b46c. Initial automatic compaction now receives the resolved run limits and a zero run-usage baseline, and checks request, token, and cost follow-up capacity before summary generation. Between-tool compaction performs the same check against live run usage and marks a reached limit instead of spending another request. The end-to-end regression forces compaction with max_requests = 0 and verifies zero summary calls and zero request usage.

File: R/agent.R [hunk 24]
```diff
@@ -350,76 +351,324 @@ Agent <- R6::R6Class(
...
+    get_system_prompt = function() {
+      private$.chat$get_system_prompt()
+    },
+
+    #' @description Replace the system prompt, as in ellmer Chat.
+    #' @param value The new system prompt or `NULL`.
+    #' @return Invisible self.
+    set_system_prompt = function(value) {
+      previous_summary <- private$.compaction_summary
+      private$.chat$set_system_prompt(value)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R:576]: Fixed in e7d36c6. Public set_system_prompt now clears the hook-context hash cache after the backend accepts the replacement. A repeated hook payload can therefore be appended to the new prompt instead of being suppressed by stale de-duplication state. The regression appends context, replaces the prompt, repeats the same context, and verifies it is present with a fresh single hash.

File: R/agent.R [hunk 25]
```diff
@@ -2968,648 +3744,224 @@ Agent <- R6::R6Class(
...
+      }
+      start_marker <- "\n\n## Previous Conversation Summary\n"
+      end_marker <- "\n\n## End Previous Conversation Summary"
+      start <- regexpr(start_marker, prompt, fixed = TRUE)
+      if (start[[1L]] < 0L) {
+        return(NULL)
+      }
+      summary_start <- start[[1L]] + attr(start, "match.length")
+      remainder <- substr(prompt, summary_start, nchar(prompt))
+      end <- regexpr(end_marker, remainder, fixed = TRUE)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R]: Fixed in e7d36c6. Deputy compaction blocks now carry a character length and SHA-256 framing header plus a package-owned end marker. Parsing reads exactly the declared summary length and validates its hash before accepting the end boundary, so delimiter text inside a summary cannot terminate the block. The regression replaces and clears a summary containing the exact old end delimiter.

File: R/agents-multi.R [hunk 26]
```diff
@@ -495,36 +499,41 @@ LeadAgent <- R6::R6Class(
...
-        base <- substr(full_prompt, 1, marker_pos - 1)
-        # Trim trailing whitespace
-        base <- sub("\\s+$", "", base)
-        if (nchar(base) == 0) {
-          return(NULL)
-        }
-        return(base)
+      start_marker <- "# Available Sub-Agents"
+      end_marker <- "# End Available Sub-Agents"
+      start <- regexpr(start_marker, full_prompt, fixed = TRUE)
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agents-multi.R:526]: Fixed in e7d36c6. LeadAgent routing is now wrapped in package-owned HTML boundary markers, and registration replaces only the marked generated block. Ordinary user-authored Available Sub-Agents headings remain untouched. The regression begins with a matching user heading pair, updates the private registry, and verifies both user content and the single generated routing section survive correctly.

File: R/agent.R [hunk 27]
```diff
@@ -2063,16 +2830,33 @@ Agent <- R6::R6Class(
...
       tryCatch(
         {
-          self$chat$set_turns(session$turns)
-          self$chat$set_system_prompt(session$system_prompt)
+          imported_tool_results <- import_tool_result_envelopes(
+            restored_tool_results,
+            policy = private$.context_policy,
+            source_session_id = metadata$session_id,
+            target_session_id = private$.session_id
+          )
```
[the remark this answers is not in the document]
Comment (JamesHWade) [reply R/agent.R]: Fixed in e574f24. load_session() now stages the incoming envelope set and swaps the receiver session directory, retaining the previous directory as a rollback until chat restoration succeeds. A successful load removes stale envelopes, including when the loaded session has none; a failed load restores the receiver previous result set. Added regression coverage for replacement, empty-session clearing, round-trip saves, and rollback.
